lightning/src/ln/channelmanager.rs

```diff
@@ -4019,6 +4019,10 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 					pending_events.push(events::Event::PendingHTLCsForwardable {
 						time_forwardable: time
 					});
+				} else {
+					let mut pending_events = self.pending_events.lock().unwrap();
+					pending_events.push(events::Event::PaymentForwardedFailed {
+						source_channel_id: short_channel_id, sink_node_id: self.get_our_node_id() });
```

Comment:
I don't think the `sink` node here is always us, more likely some other node that failed back to us. I think in practice we'll need to add a new argument to `fail_htlc_backwards_internal` and fill it in at all the callsites to give us knowledge of the sink node.

Separately, note that you can get the source channel_id from the `prev_funding_outpoint` in the `HTLCPreviousHopData`, which may be more useful than the SCID for users.

Reply:
actually, I have a question about this:

if we're matching `HTLCSource::PreviousHopData` to run this block to send the `PaymentForwardingFailed` event, can we derive the sink node from `PreviousHopData`'s `short_channel_id` field?

i.e. 

```
let sink_node_id = match channel_state_lock.short_to_id.get(&short_channel_id) {
	Some(chan_id)=> channel_state_lock.by_id(chan_id).get().get_counterparty_node_id()
}
```

Reply:
at the branch's current state, I've added `receiver` as an optional parameter, and have done most of the digging around for the `counterparty_node_id`, except I got stuck for a couple I can use some help with:

1. Specifically [here when we want to fail MPP HTLCs](https://github.com/lightningdevkit/rust-lightning/pull/1403/files#diff-645d12c7269d09caab57d2e0c1fd34f672ee259994c86ad4bad7f4da9183671cL3730)
2. The public function for [failing the HTLC backwards](https://github.com/lightningdevkit/rust-lightning/pull/1403/files#diff-645d12c7269d09caab57d2e0c1fd34f672ee259994c86ad4bad7f4da9183671cR3761). The only thing that seems to give us a lead for deriving the `counterparty_node_id` was the `HTLCSource`, which was why I was wondering about the question above.
3. [Failing all HTLCs in MPP if channel doesn't exist for one](https://github.com/lightningdevkit/rust-lightning/pull/1403/files#diff-645d12c7269d09caab57d2e0c1fd34f672ee259994c86ad4bad7f4da9183671cR4098-R4099)
4. [Failing HTLCs with no `payment_preimage` when processing pending monitor events](https://github.com/lightningdevkit/rust-lightning/pull/1403/files#diff-645d12c7269d09caab57d2e0c1fd34f672ee259994c86ad4bad7f4da9183671cR5099)

Reply:
Those are all "payments to us" which maybe tell us the `PaymentForwardingFailed` event is a misnomer - its really `HTLCFailed` and can be both a forward or a to-us HTLC.